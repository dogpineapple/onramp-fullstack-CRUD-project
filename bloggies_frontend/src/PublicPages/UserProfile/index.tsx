import React, { useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { useParams } from "react-router";
import { BASE_URL } from "../../config";
import { Post } from "../../custom";
import FavoritesList from "../../FavoritesList";
import "./UserProfile.css";

function UserProfile() {
  const currUserId = useSelector((st: any) => st.user.id); 
  const { userId, displayName } = useParams<{ userId: string, displayName: string }>();
  const [ isCurrUserProfile , setIsCurrUserProfile ] = useState<boolean>(false);
  const currUserFavs = useSelector((st: any) => st.favorites);
  const [ userFavs, setUserFavs ] = useState<Array<Post>>([]);

  useEffect(function checkProfileOwner() {
    async function getAltUserFavorites() {
      const res = await fetch(`${BASE_URL}/favorites/${userId}`);
      const resData = await res.json();
      setUserFavs(resData.posts);
    }

    // if profile belongs to the current user, use redux data.
    if (parseInt(userId) === currUserId) {
      setIsCurrUserProfile(true);
    } else { 
      // if not current user's profile, fetch the data from backend.
      getAltUserFavorites();
    }
  }, []);

  return (
    <div className="UserProfile">
      <h1 className="mt-4">{displayName.replaceAll("-", " ")}'s profile</h1>
      <FavoritesList displayName={displayName} favorites={isCurrUserProfile ? currUserFavs : userFavs}/>
    </div>
  );
};

export default UserProfile;