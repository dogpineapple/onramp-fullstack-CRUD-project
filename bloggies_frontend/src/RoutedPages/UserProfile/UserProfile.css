.UserProfile-photo {
  margin: 1.5rem 0rem;
}