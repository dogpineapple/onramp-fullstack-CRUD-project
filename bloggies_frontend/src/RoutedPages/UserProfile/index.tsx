import React, { useEffect, useState } from "react";
import { Container, Row, Col, Alert } from "react-bootstrap";
import { useSelector } from "react-redux";
import { useParams } from "react-router";
import BlogList from "../../BlogList";
import { BASE_URL } from "../../config";
import { CustomReduxState, Post } from "../../custom";
import FavoritesList from "../../FavoritesList";
import "./UserProfile.css";

/**
 * `UserProfile` renders the page for displaying a user's `FavoritesList` and 
 * Publications (`BlogList`).
 */
function UserProfile() {
  const currUserId = useSelector((st: CustomReduxState) => st.user.id);
  const { userId, displayName } = useParams<{ userId: string, displayName: string }>();
  const [isCurrUserProfile, setIsCurrUserProfile] = useState<boolean>(false);
  const currUserFavs = useSelector((st: CustomReduxState) => st.favorites);
  const [userFavs, setUserFavs] = useState<Array<Post>>([]);
  const [userPosts, setUserPosts] = useState<Array<Post>>([]);
  const [serverErr, setServerErr] = useState("");

  useEffect(function checkProfileOwner() {
    // retrieve user favorites by a GET request with user id from params.
    async function getUserFavorites() {
      try {
        const favRes = await fetch(`${BASE_URL}/favorites/${userId}`);
        const favData = await favRes.json();
        setUserFavs(favData.posts);
      } catch (err) {
        setServerErr("This user does not exist.");
      }
    }

    // retrieve user publications by a GET request with user id from params.
    async function getUserPosts() {
      try {
        const userPostsRes = await fetch(`${BASE_URL}/posts/user/${userId}`);
        const userPostsData = await userPostsRes.json();
        setUserPosts(userPostsData.posts);
      } catch (err) {
        setServerErr("This user does not exist.");
      }
    }

    // if profile belongs to the current user, use redux data.
    if (parseInt(userId) === currUserId) {
      setIsCurrUserProfile(true);
      getUserPosts();
    } else {
      // if not current user's profile, fetch the data from backend.
      getUserFavorites();
      getUserPosts();
    }
  }, [userId]);

  return (
    <Container className="UserProfile">
      <h1 className="mt-4">{displayName.replaceAll("-", " ")}'s profile</h1>
      { serverErr && <Alert>{serverErr}</Alert>}
      <Row>
        <Col md={6}>
          <Container>
            <Row>
              <h3 className="mt-4">Publications</h3>
            </Row>
            <Row>
              <BlogList posts={userPosts} />
            </Row>
          </Container>
        </Col>
        <Col md={6}>
          <FavoritesList favorites={isCurrUserProfile ? currUserFavs : userFavs} />
        </Col>
      </Row>
    </Container>
  );
};

export default UserProfile;