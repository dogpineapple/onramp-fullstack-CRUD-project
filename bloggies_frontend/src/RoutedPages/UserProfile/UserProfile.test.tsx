import React, { Component } from 'react';
import Enzyme, { mount, ReactWrapper, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import UserProfile from './index';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import ReactRouter from 'react-router';
import { MOCK_STORE } from '../../jest.mock';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();


describe('SearchBar', () => {
  let wrapper: ReactWrapper;

  beforeEach(() => {
    jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ userId: '1', displayName: 'test-user' });
 
    wrapper = mount(shallow<Component>(
      <Provider store={mockStore(MOCK_STORE)}>
        <UserProfile />
      </Provider>
    ).get(0));
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should display user name', () => {
    const displayNameTag = wrapper.find(".UserProfile > h1");
    expect(displayNameTag.text()).toBe('test user\'s profile');
  });
});