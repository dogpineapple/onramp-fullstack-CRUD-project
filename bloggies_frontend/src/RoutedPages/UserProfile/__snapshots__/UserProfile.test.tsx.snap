// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match snapshot 1`] = `
<UserProfile>
  <Container
    className="UserProfile"
    fluid={false}
  >
    <div
      className="UserProfile container"
    >
      <h1
        className="mt-4"
      >
        test user
        's profile
      </h1>
      <Row
        noGutters={false}
      >
        <div
          className="row"
        >
          <Col
            md={6}
          >
            <div
              className="col-md-6"
            >
              <Container
                fluid={false}
              >
                <div
                  className="container"
                >
                  <Row
                    noGutters={false}
                  >
                    <div
                      className="row"
                    >
                      <h3
                        className="mt-4"
                      >
                        Publications
                      </h3>
                    </div>
                  </Row>
                  <Row
                    noGutters={false}
                  >
                    <div
                      className="row"
                    >
                      <BlogList
                        posts={Array []}
                      >
                        <Container
                          className="BlogList"
                          fluid={false}
                        >
                          <div
                            className="BlogList container"
                          >
                            <Row
                              noGutters={false}
                            >
                              <div
                                className="row"
                              >
                                <Col
                                  md={12}
                                >
                                  <div
                                    className="col-md-12"
                                  >
                                    <p
                                      className="mt-5"
                                    >
                                      No posts made yet.
                                    </p>
                                  </div>
                                </Col>
                              </div>
                            </Row>
                          </div>
                        </Container>
                      </BlogList>
                    </div>
                  </Row>
                </div>
              </Container>
            </div>
          </Col>
          <Col
            md={6}
          >
            <div
              className="col-md-6"
            >
              <FavoritesList
                favorites={Array []}
              >
                <Container
                  className="FavoritesList"
                  fluid={false}
                >
                  <div
                    className="FavoritesList container"
                  >
                    <Row
                      noGutters={false}
                    >
                      <div
                        className="row"
                      >
                        <h3
                          className="mt-4"
                        >
                          Favorites
                        </h3>
                      </div>
                    </Row>
                    <Row
                      noGutters={false}
                    >
                      <div
                        className="row"
                      >
                        <div>
                          Empty... for now!
                        </div>
                      </div>
                    </Row>
                  </div>
                </Container>
              </FavoritesList>
            </div>
          </Col>
        </div>
      </Row>
    </div>
  </Container>
</UserProfile>
`;
